src/Hedgehog/Gen.fs

```diff
@@ -247,26 +248,32 @@ module Gen =
     /// More or less the same logic as suchThatMaybe from QuickCheck, except
     /// modified to ensure that the shrinks also obey the predicate.
     let private tryFilterRandom (p : 'a -> bool) (r0 : Random<Tree<'a>>) : Random<Option<Tree<'a>>> =
```

Comment:
I am having a hard time following changes in functions like this one.  (See main review comment for more information.)